File: src/main/java/hu/frontrider/arcana/content/items/formula/text/EnchantText.java
package hu.frontrider.arcana.content.items.formula.text;

import hu.frontrider.arcana.core.creatureenchant.CreatureEnchant;
import hu.frontrider.arcana.util.AspectUtil;
import hu.frontrider.arcana.util.StringUtil;
import net.minecraft.client.resources.I18n;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.registry.GameRegistry;
import net.minecraftforge.registries.IForgeRegistry;
import thaumcraft.api.aspects.AspectList;

import javax.annotation.Nullable;
import java.util.List;

/**
 * @author Kis András Gábor
 * 2018.08.28.
 */
public class EnchantText implements FormulaText {
    private IForgeRegistry<CreatureEnchant> creatureEnchantIForgeRegistry;

    public EnchantText() {
        creatureEnchantIForgeRegistry = GameRegistry.findRegistry(CreatureEnchant.class);
    }

    @Override
    public void addText(ItemStack stack, NBTTagCompound tagCompound, @Nullable World worldIn, List<String> tooltip) {
        AspectList aspectList = new AspectList();
        aspectList.readFromNBT(tagCompound);
        StringUtil.INSTANCE.swappedDescription(tooltip)
                .permanent(description -> creatureEnchantIForgeRegistry.getValuesCollection().forEach(creatureEnchant -> {
                    if (AspectUtil.INSTANCE.aspectListEquals(aspectList, creatureEnchant.formula())) {
                        description.add(I18n.format("enchant.creature_enchant.helper.enchant") + ": " + I18n.format(creatureEnchant.getUnlocalizedName()));
                    }
                }))
                .render();
    }
}

File: src/main/java/hu/frontrider/arcana/sided/client/gui/formula/text/EnchantText.kt
package hu.frontrider.arcana.sided.client.gui.formula.text

import hu.frontrider.arcana.core.creatureenchant.CreatureEnchant
import hu.frontrider.arcana.util.AspectUtil
import hu.frontrider.arcana.util.strings.ChatFormat.YELLOW
import hu.frontrider.arcana.util.strings.formatTranslate
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.world.World
import net.minecraftforge.fml.common.registry.GameRegistry
import thaumcraft.api.aspects.AspectList

/**
 * @author Kis András Gábor
 * 2018.08.28.
 */
class EnchantText : FormulaText {

    private val creatureEnchantIForgeRegistry = GameRegistry.findRegistry(CreatureEnchant::class.java)

    override fun addText(stack: ItemStack, tagCompound: NBTTagCompound, worldIn: World?, tooltip: MutableList<String>) {
        val aspectList = AspectList()
        aspectList.readFromNBT(tagCompound)

        creatureEnchantIForgeRegistry.valuesCollection.forEach { creatureEnchant ->
            if (AspectUtil.aspectListEquals(aspectList, creatureEnchant.formula())) {
                tooltip+=("${formatTranslate("enchant.creature_enchant.helper.enchant", YELLOW)}: ${formatTranslate(creatureEnchant.getUnlocalizedName(), YELLOW)}")
            }
        }
    }
}
